test/UtilsTest.js

```diff
@@ -244,212 +245,124 @@ describe('FlowLimit', async () => {
-        await test.addFlowIn(1).then((tx) => tx.wait);
-        await test.addFlowOut(limit).then((tx) => tx.wait);
-
-        await expectRevert((gasOptions) => test.addFlowOut(2, gasOptions), test, 'FlowLimitExceeded', [limit, limit.add(1), test.address]);
-    });
-
-    it('Should revert when in/out/in combination creates net imbalance over the limit', async () => {
-        const half = MaxUint256.div(2);
-        const limit = half;
+        if (error === 'FlowAmountOverflow') {
+            const current = isIn ? await test.flowInAmount() : await test.flowOutAmount();
+            return [amount, BigNumber.from(current), addr];
+        } else if (error === 'FlowLimitExceeded') {
+            const forwardFlow = isIn ? flowIn.add(amount) : flowOut.add(amount);
+            const reverseFlow = isIn ? flowOut : flowIn;
+            const netFlow = forwardFlow.gt(reverseFlow) ? forwardFlow.sub(reverseFlow) : reverseFlow.sub(forwardFlow);
+            return [flowLimit, netFlow, addr];
+        }
 
-        await test.setFlowLimit(limit).then((tx) => tx.wait);
-        await test.addFlowIn(half).then((tx) => tx.wait);
-        await test.addFlowOut(1).then((tx) => tx.wait);
+        // Default case: FlowAmountExceededLimit
+        return [flowLimit, amount, addr];
+    };
+
+    const executeTestCases = (error, testCases) => {
+        const allCases = [
+            ...testCases,
+            ...testCases.map(({ flowLimit, flows }) => ({
+                flowLimit,
+                flows: flipFlows(flows),
+            })),
+        ];
 
-        await expectRevert((gasOptions) => test.addFlowIn(2, gasOptions), test, 'FlowLimitExceeded', [limit, half.add(1), test.address]);
-    });
+        for (const { flowLimit, flows } of allCases) {
+            const desc = `Should revert with ${error} (flowLimit=${flowLimit}, steps=${flows.length})`;
+
+            it(desc, async () => {
+                await nextEpoch();
+                await test.setFlowLimit(flowLimit).then((tx) => tx.wait);
+
+                let flowIn = BigNumber.from(0);
+                let flowOut = BigNumber.from(0);
+
+                for (let i = 0; i < flows.length; i++) {
+                    const { in: inAmount, out: outAmount } = flows[i];
+                    const isIn = inAmount != null;
+                    const amount = BigNumber.from(inAmount ?? outAmount);
+                    const fn = isIn ? test.addFlowIn : test.addFlowOut;
+                    const isLast = i === flows.length - 1;
+
+                    if (isLast) {
+                        const expectedArgs = await calculateExpectedArgs(error, isIn, amount, flowIn, flowOut, flowLimit, test);
+                        await expectRevert((gasOptions) => fn(amount, gasOptions), test, error, expectedArgs);
+                    } else {
+                        await fn(amount).then((tx) => tx.wait);
+                        isIn ? (flowIn = flowIn.add(amount)) : (flowOut = flowOut.add(amount));
```

Comment:
bad practice to use side effects in the ternary operator. also not very readable